libs/community/tests/unit_tests/vectorstores/test_sqlserver.py

```diff
@@ -594,6 +594,37 @@ def test_that_entra_id_authentication_connection_is_successful(
     vector_store.drop()
 
 
+def test_that_max_marginal_relevance_search_returns_expected_no_of_documents(
+    store: SQLServer_VectorStore,
+    texts: List[str],
+) -> None:
+    """Test that the size of documents returned when `max_marginal_relevance_search`
+    is called is the expected number of documents requested."""
+    store.add_texts(texts)
+    number_of_docs_to_return = 3
+    result = store.max_marginal_relevance_search(
+        query="Good review", k=number_of_docs_to_return
+    )
+    assert len(result) == number_of_docs_to_return
+
+
+def test_similarity_search_with_relevance_score(
+    store: SQLServer_VectorStore,
+    texts: List[str],
+    metadatas: List[dict],
+) -> None:
+    """Test that the size of documents returned when
+    `similarity_search_with_relevance_scores` is called
+    is the expected number of documents requested."""
+    number_of_docs_to_return = 3
+
+    store.add_texts(texts, metadatas)
+    result = store.similarity_search_with_relevance_scores(
+        "Good review", k=number_of_docs_to_return
+    )
+    assert len(result) == number_of_docs_to_return
```

Comment:
this is only testing we get the expected number of docs. Can we also check the result is constrained between 0 and 1. 

Also, is there any way to test if the results we are getting are as expected as per maximal_marginal_relevance? Meaning that those would be different if we were to NOT use such a relevance_score?

Reply:
Our results here aren't always constrained between 0 and 1. That's what @yorek will be looking into.
If we do this, we'll be introducing flakiness to this test.

Yes, we can determine expected value with the formula -> (1 -`similarity_search_score`). And check that the result from the call to `similarity_search_with_relevance_scores` is expected.

Reply:
I checked the test, and the issue is the vector are randomly generated by the `FakeEmbeddings` function. I tried using vectors with only three dimensions and the generated vectors where:

v1 = [-2.3745856e+000,1.2014049e+000,6.8701744e-001]
v2 = [5.9738564e-001,3.5180563e-001,4.1327056e-001]

if you calculate the cosine distance the result is 1.32094580014782 which means that the resulting similarity (or score) is `1-1.32094580014782` = -0.32094580014782

Here's the manual proof:

```sql
with 
v1 as (select cast(value as float) as v, * from openjson('[-2.3745856e+000,1.2014049e+000,6.8701744e-001]')),
v2 as (select cast(value as float) as v,* from openjson('[5.9738564e-001,3.5180563e-001,4.1327056e-001]'))
select
 1-SUM(v1.v * v2.v) / 
    (
        SQRT(SUM(v1.v * v1.v)) 
        * 
        SQRT(SUM(v2.v * v2.v))
    ) as cosine_distance
from
v1 inner join v2 on v1.[key] = v2.[key]
```

I think this issue can be ignored for now. I would suggest opening an issue on the langchain repo to ask for confirmation and/or propose to use only positive values for generating vectors, after making that change, test was running green all the time, while before it was very inconsistent (as expected given the vectors are completely random)/

This is the change I did in `fake.py` file and added `abs` going from

```python
    def _get_embedding(self) -> List[float]:
        return list(np.random.normal(size=self.size))
```

to

```python
    def _get_embedding(self) -> List[float]:
        return list(abs(np.random.normal(size=self.size)))
```

to make sure generated vectors are always positive and the cosine distance is between 0 and 1

Reply:
Actually, I submitted a PR already :) https://github.com/langchain-ai/langchain/pull/27406